examples/model_compress/pruning/taylorfo_pruning_with_ddp_torch.py

```diff
@@ -0,0 +1,257 @@
+        lr = args.pretraining_lr
+    else: # finetune process
+        log_save_path = "finetune.log"
+        model_save_path = "finetune_best_model.pth"
+        epochs = args.finetune_epochs
+        lr = args.finetune_lr
+
+    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.local_rank])
+
+    # create dataloaders
+    training_dataloader, validation_dataloader = create_dataloaders()
+    # create optimizer, lr_scheduler and criterion
+    optimizer, lr_scheduler = optimizer_scheduler_generator(model, \
+        _lr=lr, total_epoch=epochs)
+    criterion = torch.nn.CrossEntropyLoss()
+
+    # training and evaluation process
+    evaluation_func = functools.partial(evaluation, validation_dataloader)
+    best_acc = training(training_dataloader, model, optimizer, criterion, lr_scheduler,\
+        args.max_steps, epochs, args.local_rank, save_best_model=True, \
+        save_path=Path(args.log_dir) / model_save_path, \
+        log_path=Path(args.log_dir) / log_save_path, \
+        evaluation_func=evaluation_func)
+
+    # compute params and FLOPs
+    flops, params, _ = count_flops_params(model, torch.randn([32, 1, 28, 28]).cuda())
+
+    return flops, params, best_acc
+
+def pruned_model_process(args):
+    # load the pretrained model
+    model = Mnist().cuda()
+    state_dict = torch.load(Path(args.log_dir) / "pretraining_best_model.pth")
+    model.load_state_dict(state_dict)
+    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[args.local_rank])
+    # create dataloaders
+    training_dataloader, validation_dataloader = create_dataloaders()
+    # build a config_list
+    config_list = [{'total_sparsity': 0.7, 'op_types': ['Conv2d']}]
+    # create an evaluator
+    taylor_training = functools.partial(
+        training,
+        training_dataloader,
+        local_rank = args.local_rank,
+        log_path = Path(args.log_dir) / "taylor_pruning.log",
+        evaluation_func = functools.partial(evaluation, validation_dataloader),
+    )
+    traced_optimizer = nni.trace(torch.optim.SGD)(model.parameters(), lr=0.01, momentum=0.9, weight_decay=5e-4)
+    criterion = torch.nn.CrossEntropyLoss()
+    evaluator = TorchEvaluator(taylor_training, traced_optimizer, criterion)
+    # create an taylor pruner
+    pruner = TaylorFOWeightPruner(model=model, config_list=config_list,
+                evaluator=evaluator, training_steps=args.pruner_training_steps)
+    _, masks = pruner.compress()
+    pruner.show_pruned_weights()
+    pruner._unwrap_model()
+    #speedup
+    sub_module = ModelSpeedup(model, dummy_input=torch.rand([32, 1, 28, 28]).cuda(), masks_file=masks).speedup_model()
+
+    return sub_module
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='PyTorch Example for model comporession with ddp')
+    parser.add_argument('--local_rank', type=int, default=-1,
+                        help='local_rank is needed when you use ddp mode')
```

Comment:
could we use this instead of pass `local_rank` as args?
```
rank = dist.get_rank()
local_rank = rank % torch.cuda.device_count()
```

Reply:
good suggestion, I will apply it.